### DISCUSSION TITLE
[Remix] Using per-page edge functions with incompatible node modules

### DISCUSSION THREAD

**Discussion by WesleyYue** [vercel/vercel#9537] [Help] [answered]
Was trying to adopt this [new feature](https://github.com/vercel/vercel/pull/9375), but I get an error like the following:
```
Error: The Edge Function "404" is referencing unsupported modules:
@remix-run: crypto, fs, path, stream, fs/promises, os, util, http, https, zlib, url
...
```

What I did:
1. Adopted the @vercel/remix-entry-server entry.server files with my modifications (see below, I don't think this is the issue)
2. Set VERCEL_CLI_VERSION = https://vercel-git-update-remix-edge.vercel.sh/tarballs/vercel.tgz
3. ~Deployed to Vercel and verify that my changes work without any edge functions (only node/lambda serverless functions)~ See update below
4. Create a new barebones edge route copy and pasting from the [example](https://github.com/vercel/vercel/blob/main/examples/remix/app/routes/edge.tsx)
5. Re-deploying to vercel
6. I get the above error on the new deployment

My guess on the issue:
I am assuming 404 is some 404 page that Vercel is trying to add on, since I don't have any routes that use the text "404", and the deployment error switches between 404, and "edge" (which I am assuming is from the example edge function, named "edge") when I redeploy the same code. In this process, I am assuming it's trying to bundle all the dependencies in my package, and this of course cannot be deployed to cf workers as the packages rely on the node api.

My question:
Am I missing a step? How can I tell Vercel to only bundle the dependencies related to the routes I mark as runtime=edge?

@TooTallNate I'm assuming you're the one that knows the answer. Thanks for shipping this, very excited to try to get it working.

My changes to the entry.server files

```typescript
// entry.server.tsx (in my remix package)
export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const emotionCache = createEmotionCache();
  const emotionServer = createEmotionServer(emotionCache);

  const html = renderToString(
    <EmotionCacheProvider value={emotionCache}>
      <RemixServer context={remixContext} url={request.url} />
    </EmotionCacheProvider>
  );

  const emotionChunks = emotionServer.extractCriticalToChunks(html);
  const emotionCss = emotionServer.constructStyleTagsFromChunks(emotionChunks);

  // swap out default component with <Head>
  const defaultRoot = remixContext.routeModules.root;
  remixContext.routeModules.root = {
    ...defaultRoot,
    default: Head,
  };

  const head = renderToString(<RemixServer context={remixContext} url={request.url} />);

  // restore the default root component
  remixContext.routeModules.root = defaultRoot;

  const remixComponent = (
    <EmotionCacheProvider value={emotionCache}>
      <RemixServer context={remixContext} url={request.url} />
    </EmotionCacheProvider>
  );

  const htmlStart = `<!DOCTYPE html><html lang="en"><head><!--start head-->${head}<!--end head-->${emotionCss}</head><body><div id="root">`;
  const htmlEnd = "</div></body></html>";

  return vercelHandleRequest(
    request,
    responseStatusCode,
    responseHeaders,
    remixComponent,
    htmlStart,
    htmlEnd
  );
}
```

```typescript
// entry.server.node.ts (in a separate, custom, remix-entry-server package in my monorepo)

const ABORT_DELAY = 5000;

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixServer: JSX.Element,
  htmlStart: string,
  htmlEnd: string
) {
  // If the request is from a bot, we want to wait for the full
  // response to render before sending it to the client. This
  // ensures that bots can see the full page content.
  if (isbot(request.headers.get("user-agent"))) {
    return serveTheBots(responseStatusCode, responseHeaders, remixServer, htmlStart, htmlEnd);
  }

  return serveBrowsers(responseStatusCode, responseHeaders, remixServer, htmlStart, htmlEnd);
}

function serveTheBots(
  responseStatusCode: number,
  responseHeaders: Headers,
  remixServer: any,
  htmlStart: string,
  htmlEnd: string
) {
  return new Promise((resolve, reject) => {
    const { pipe, abort } = renderToPipeableStream(remixServer, {
      // Use onAllReady to wait for the entire document to be ready
      onAllReady() {
        responseHeaders.set("Content-Type", "text/html");

        const htmlStream = new PassThrough();

        htmlStream.write(htmlStart);
        pipe(htmlStream);
        htmlStream.end(htmlEnd);

        resolve(
          new Response(htmlStream, {
            status: responseStatusCode,
            headers: responseHeaders,
          })
        );
      },
      onShellError(err) {
        reject(err);
      },
    });
    setTimeout(abort, ABORT_DELAY);
  });
}

function serveBrowsers(
  responseStatusCode: number,
  responseHeaders: Headers,
  remixServer: any,
  htmlStart: string,
  htmlEnd: string
) {
  return new Promise((resolve, reject) => {
    let didError = false;
    const { pipe, abort } = renderToPipeableStream(remixServer, {
      // use onShellReady to wait until a suspense boundary is triggered
      onShellReady() {
        responseHeaders.set("Content-Type", "text/html");

        const htmlStream = new PassThrough();

        htmlStream.write(htmlStart);
        pipe(htmlStream);
        htmlStream.end(htmlEnd);
        resolve(
          new Response(htmlStream, {
            status: didError ? 500 : responseStatusCode,
            headers: responseHeaders,
          })
        );
      },
      onShellError(err) {
        reject(err);
      },
      onError(err) {
        didError = true;
        console.error(err);
      },
    });
    setTimeout(abort, ABORT_DELAY);
  });
}
```

```typescript
// entry.server.edge.ts (in a separate, custom, remix-entry-server package in my monorepo)
// Modified from https://github.com/remix-run/remix/discussions/5244#discussioncomment-4890083

import isbot from "isbot";
import { renderToReadableStream } from "react-dom/server";

const readableString = (value: string) => {
  const te = new TextEncoder();
  return new ReadableStream({
    start(controller) {
      controller.enqueue(te.encode(value));
      controller.close();
    },
  });
};

export default async function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixServer: JSX.Element,
  htmlStart: string,
  htmlEnd: string
) {
  const { readable, writable } = new TransformStream();
  const head = readableString(htmlStart);
  const end = readableString(htmlEnd);

  const body = await renderToReadableStream(remixServer, {
    signal: request.signal,
    onError(error) {
      console.error(error);
      responseStatusCode = 500;
    },
  });

  Promise.resolve()
    .then(() => head.pipeTo(writable, { preventClose: true }))
    .then(() => body.pipeTo(writable, { preventClose: true }))
    .then(() => end.pipeTo(writable));

  if (isbot(request.headers.get("user-agent"))) {
    await body.allReady;
  }

  responseHeaders.set("Content-Type", "text/html");
  return new Response(readable, {
    headers: responseHeaders,
    status: responseStatusCode,
  });
}
```

**Comment by WesleyYue** [comment]
Actually, I lied, the successful deploy (with only node/lambda functions) is without setting VERCEL_CLI_VERSION = [vercel-git-update-remix-edge.vercel.sh/tarballs/vercel.tgz](https://vercel-git-update-remix-edge.vercel.sh/tarballs/vercel.tgz). When I set the new VERCEL_CLI_VERSION and only deploy node/lambda serverless functions, I get this:

![image](https://user-images.githubusercontent.com/5280818/221116467-4c465226-ea0a-4cc8-99eb-1eb2abd53ac6.png)

And there's nothing informative in the server logs, just a 500 response with the regular vercel response headers.

**Comment by the-artifabrian** [comment]
I just got the same error when trying out the edge mode, did you manage to get around it @WesleyYue?

**Reply by WesleyYue** [reply]
I think they are still working on it: https://github.com/remix-run/remix/pull/5479

**Reply by WesleyYue** [reply]
The only way to use edge right now is if your entire package is edge compatible I think

**Comment by leerob** [comment]
We're working on making Remix multi-runtime, to have granularity between Node.js / Edge on a per-route basis!

**Comment by TooTallNate** [comment, marked as answer]
In short, this issue happens because you're importing `@remix-run/node` in a page using Edge runtime. If you're just trying to import some of the generic helper functions such as `json()`, `defer()`, or a TypeScript type, these are available on other Edge-compatible runtime packages, such as `@remix-run/cloudflare`.

Ideally you won't have to use `cloudflare` since it's more of an implementation detail, and we're working on the story there. Should get you unblocked for now though.

**Reply by WesleyYue** [reply]
Makes sense, thanks for all of your work on this Nate! A few questions:

- Should I expect it to work once I update the import to @remix-run/cloudflare? Or do we need to wait for [this PR](https://github.com/remix-run/remix/pull/5479)?
- What is the correct strategy for common components shared across different routes that use non-edge-compatible packages? Or do we nee to make two versions of the component?

**Comment by leerob** [comment]
Good news! https://twitter.com/vercel/status/1638656244566327298

**Comment by cill-i-am** [comment]
@TooTallNate i'm having this issue at the moment and I'm using @vercel/remix. Don't have any references to @remix-run/node in the project

**Reply by TooTallNate** [reply]
Can you share more details (ideally with a minimal reproduction) in a new discussion thread?

**Reply by cill-i-am** [reply]
Hey! Same as below. Had an unsupported package further up the tree that was breaking it. thanks for the quick response though

**Comment by bill-kerr** [comment]
I am also having this issue despite not having `@remix-run/node` installed in my project.

**Reply by TooTallNate** [reply]
Can you share more details (ideally with a minimal reproduction) in a new discussion thread?

**Reply by bill-kerr** [reply]
It was my mistake. I had imported unsupported packages in `root.tsx`, but the error was a bit misleading since it was pointing me toward the `404` page that I guess Vercel adds?

**Reply by bill-kerr** [reply]
Thanks for following up though

**Reply by keepyara** [reply]
@bill-kerr I am having the same issue, I get the error referencing 404, and I also don't have any references to a `@remix-run/node`. What unsupported packages were you using?

**Comment by Kikketer** [comment]
Seeing the same issue, bare minimum way to reproduce this:

1. Create the app with the Remix Boilerplate example
2. Run `npx remix reveal`
3. Push and witness the errors

Attempting to change to the `cloudflare` version also didn't help

The baseline repo that fails:
https://github.com/Kikketer/remix-boilerplate

![image](https://github.com/vercel/vercel/assets/958042/5c71347e-56c5-4be0-9c85-669384db35f6)
